ulib/LowStar.ConstBuffer.fsti

```diff
@@ -0,0 +1,187 @@
+type qual =
+  | MUTABLE
+  | IMMUTABLE
+
+/// `qbuf_cases q a`: disjunction of mutable and immutable pointers
+let qbuf_cases (q:qual) (a:Type) =
+  match q with
+  | MUTABLE -> B.buffer a
+  | IMMUTABLE -> I.ibuffer a
+
+/// `q_preorder q a`: As we'll see shortly, it is convenient to also
+/// define a disjunction of the preorder indices on a qualified
+/// buffer
+let q_preorder (q:qual) (a:Type) =
+  match q with
+  | MUTABLE -> B.trivial_preorder a
+  | IMMUTABLE -> I.immutable_preorder a
+
+/// `qbuf a`: This type is used for specificational purposes only. It
+///  is buffer whose mutability qualifier is existentially bound, via
+///  a dependent pair
+let qbuf a = (q:qual & qbuf_cases q a)
+
+/// `qbuf_qual c`: projecting the mutability qualifier of a `qbuf`
+let qbuf_qual (c:qbuf 'a) : qual = dfst c
+
+/// `qbuf_pre c`: case-dependent preorders for a qbuf
+let qbuf_pre (c:qbuf 'a)  = q_preorder (qbuf_qual c) 'a
+
+/// `qbuf_mbuf c`: turning a qbuf into a regular monotonic buffer
+let qbuf_mbuf (c:qbuf 'a) : B.mbuffer 'a (qbuf_pre c) (qbuf_pre c) = dsnd c
+
+(*** CONCRETE CONST POINTERS **)
+
+/// `const_buffer`:
+///  An abstract type of a read-only pointer to an array of `a`
+val const_buffer (a:Type u#0) : Type u#0
+
+
+/// `as_qbuf`: For specificational purposes, a const_buffer can be
+/// seen as an existentially quantified qbuf
+val as_qbuf (c:const_buffer 'a) : GTot (qbuf 'a)
+
+/// `as_mbuf`: A convenience function to turn a const_buffer into a
+/// regular mbuffer, for spec purposes
+let as_mbuf (c:const_buffer 'a) : GTot _ = qbuf_mbuf (as_qbuf c)
+
+/// We now give several convenience functions that lift common
+/// notions on buffers to const_buffer, via the `as_mbuf` coercion
+let live (h:HS.mem) (c:const_buffer 'a)   = B.live h (as_mbuf c)
+let length (c:const_buffer 'a)            = B.length (as_mbuf c)
+let loc_buffer (c:const_buffer 'a)        = B.loc_buffer (as_mbuf c)
+let as_seq (h:HS.mem) (c:const_buffer 'a) = B.as_seq h (as_mbuf c)
+
+(*** CONSTRUCTORS **)
+
+/// `of_buffer`: A constructors for const buffers from mutable and
+/// immutable buffers. It is fully specified in terms of the
+/// `qbuf/mbuf` model
+val of_buffer (b:B.buffer 'a)
+  : Pure (const_buffer 'a)
+    (requires True)
+    (ensures fun c ->
+      let c = as_qbuf c in
+      qbuf_qual c == MUTABLE /\
+      qbuf_mbuf c == b)
+
+/// `of_ibuffer`: A constructors for const buffers from mutable and
+/// immutable buffers. It is fully specified in terms of the
+/// `qbuf/mbuf` model
+val of_ibuffer (b:I.ibuffer 'a)
+  : Pure (const_buffer 'a)
+    (requires True)
+    (ensures fun c ->
+      let c = as_qbuf c in
+      qbuf_qual c == IMMUTABLE /\
+      qbuf_mbuf c == b)
+
+(*** OPERATIONS ON CONST POINTERS **)
+
+/// `index c i`: Very similar to the spec for `Buffer.index`
+val index (c:const_buffer 'a) (i:U32.t)
+  : Stack 'a
+    (requires fun h ->
+      live h c /\
+      U32.v i < length c)
+    (ensures fun h y h' ->
```

Comment:
This is again a small convention thing but usually we use `h0` and `h1` for the pre and post memories. Would be good to be consistent that way.

Reply:
nice catch. will change. Not sure why I chose `'`

Reply:
oh, cut and paste of LowStar.Monotonic.Buffer.index